lib/src/main/java/com/diffplug/spotless/DirtyState.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright 2022-2024 DiffPlug
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.diffplug.spotless;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.file.Files;
+import java.util.Arrays;
+
+import javax.annotation.Nullable;
+
+/**
+ * The clean/dirty state of a single file.  Intended use:
+ * - {@link #isClean()} means that the file is is clean, and there's nothing else to say
+ * - {@link #didNotConverge()} means that we were unable to determine a clean state
+ * - once you've tested the above conditions and you know that it's a dirty file with a converged state,
+ * then you can call {@link #writeCanonicalTo(OutputStream)} to get the canonical form of the given file.
+ */
+public class DirtyState {
+	@Nullable
+	private final byte[] canonicalBytes;
+
+	DirtyState(@Nullable byte[] canonicalBytes) {
+		this.canonicalBytes = canonicalBytes;
+	}
+
+	public boolean isClean() {
+		return this == isClean;
+	}
+
+	public boolean didNotConverge() {
+		return this == didNotConverge;
+	}
+
+	private byte[] canonicalBytes() {
+		if (canonicalBytes == null) {
+			throw new IllegalStateException("First make sure that {@code !isClean()} and {@code !didNotConverge()}");
+		}
+		return canonicalBytes;
+	}
+
+	public void writeCanonicalTo(File file) throws IOException {
+		Files.write(file.toPath(), canonicalBytes());
+	}
+
+	public void writeCanonicalTo(OutputStream out) throws IOException {
+		out.write(canonicalBytes());
+	}
+
+	/** Returns the DirtyState which corresponds to {@code isClean()}. */
+	public static DirtyState clean() {
+		return isClean;
+	}
+
+	static final DirtyState didNotConverge = new DirtyState(null);
+	static final DirtyState isClean = new DirtyState(null);
```

Comment:
```suggestion
	static final DirtyState cleanState = new DirtyState(null);
```

Reply:
reasonable suggestion, but I'm sticking with `isClean` because I like the fluent style